### ISSUE TITLE
Export/Import queries

### ISSUE THREAD

**Issue by brendam** [digitalmethodsinitiative/dmi-tcat#143]
I'd like to have an option to export the capture settings (query bins) from one DMI-TCAT server and import them into another DMI-TCAT server. Even if it had to be a clean import (rather than an 'add' to existing bins) it would still be useful for moving servers and setting up backup servers when expecting outages of main server. Just the queries, not any data that has been captured using them.

**Comment by ErikBorra** [comment]
Hi @brendam,

iirc @dentoir has implemented a full export/import feature (including data). I'm not sure whether it made its way to the master branch already. @dentoir, could you look into this? Maybe we could modify these functions to provide an option to only export/import settings.

Best,

ERik

**Comment by ErikBorra** [comment]
Hi @brendam,

we have made two scripts available via the latest code commits which will allow you to export/import capture settings with or without data. Instructions can be found here: https://github.com/digitalmethodsinitiative/dmi-tcat/wiki/FAQ#how-can-i-migrate-data-between-dmi-tcat-installations

Best,

Erik

**Comment by brendam** [comment]
Thanks, that looks great @ErikBorra, I'll give it a try.

**Comment by dentoir** [comment]
Hi @brendam 

Did you have any success with those scripts Erik mentioned?

**Comment by brendam** [comment]
Just been testing them now, sorry about the delay. 

I exported just the structure for one bucket and then deleted it (and checked in mysql that it was gone) after what seemed to be a successful import  I'm getting this error in the Apache2 error log:

`[Tue Mar 08 11:54:51.586685 2016] [:error] [pid 19221] [client 131.181.139.69:62503] PHP Fatal error:  Uncaught exception 'PDOException' with message 'SQLSTATE[42S02]: Base table or view not found: 1146 Table 'twittercapture.Test2_tweets' doesn't exist' in /var/www/dmi-tcat/analysis/common/functions.php:797\nStack trace:\n#0 /var/www/dmi-tcat/analysis/common/functions.php(797): PDOStatement->execute()\n#1 /var/www/dmi-tcat/analysis/common/functions.php(16): get_all_datasets()\n#2 /var/www/dmi-tcat/analysis/ajax.serverload.php(3): require_once('/var/www/dmi-tc...')\n#3 {main}\n  thrown in /var/www/dmi-tcat/analysis/common/functions.php on line 797, referer: http://203.101.239.132/analysis/
`

and the web pages give a 500 error.

Looking at MySQL, these all look correct:

- `tcat_query_bins` entry for Test2
- `tcat_query_bins_periods` record for Test2.
- `tcat_query_bins_phrases` records for keywords for Test2.
- `tcat_query_phrases` records for keywords for Test2

but there are none of the 'Test2_'* tables that are created if I add the bin manually. I guess the export script needs to add creating these to the sql file? I've tested this by adding the creation lines and it worked correctly. I've attached the updated sql file.
[Test2-track-queries-2016-03-07_08-42.sql.gz](https://github.com/digitalmethodsinitiative/dmi-tcat/files/162412/Test2-track-queries-2016-03-07_08-42.sql.gz)

While testing I gave the import script a .sql file instead of a .gz file and although it reported 'not in gzip format' it still said "Import completed and queries added to TCAT." when they hadn't been. I think this is the case on other errors too.

I think the export script can only do one bin at a time at the moment. How much harder would it be to have an option to export the structure of all the bins on a server?

**Comment by brendam** [comment]
I've found another problem. If you create a new bin with an already used phrase via the capture web page then it just reuses the existing phrase id. But the import creates a new phrase with a new id and deletes the original phrase without updating the phrase id in the pre-exisitng bin, so it no longer has a phrase associated with it.

**Comment by dentoir** [comment]
Hi @brendam 

Regarding your first item, if you export Test2 with the 'structure' argument you are only exporting the query phrases and are not creating the bins themselves. We may (should?) add an option to create an empty bin.

I think you should've used the 'all' argument to export both query phrases and data. I'll look into the other items now.

If you use the import script on a non-gzipped sql file, it will fail. The 'import completed' message is incorrect in this case. I've pushed a change to master to output a correct error message.

**Comment by dentoir** [comment]
Hi @brendam 

I've pushed another commit, which should fix the problem with existing phrases. The exported SQL now uses the original phrase id if a phrase should already exist in the target database. Please let us know if this corrects the problem you faced.

Regarding the use-case to fully export all bins, I'd usually suggest using the standard MySQL way of dumping an entire database.

`mysqldump --default-character-set=utf8mb4 -uuser -ppass twittercapture > alldata.sql`

But if you want to merge all bins (or just the structure/queries) into an existing database, you would have to invoke the export script multiple times (maybe script around it). The import script only does some additional safety checking to avoid importing data if a bin already exists. If you feel confident you could ignore this step and simply concatenate the .sql files and import them manually with a single command.

`zcat *.sql.gz | mysql command`

It's not impossible to edit the import/export scripts, ofcourse. It would take some refactoring.

**Comment by brendam** [comment]
I intended to just export the structure - we want to have an easy way to be able to move searches (not the data) to a new server, for example if the main server is going to be down for maintenance. The problem with import not creating the empty bin tables is that without them the capture page can't load - has an error because they don't exist.

I'll do some testing of the new version tomorrow.

**Comment by brendam** [comment]
The import of an existing phrase is working for a duplicate phrase now but has stopped working for a unique phrase. The import creates records in the tcat_query_bins, tcat_query_bins_periods, tcat_query_bins_phrases with a phrase_id of Null and nothing in the tcat_query_phrases. 

The same problem of phrase record not being created and phrase_id of Null applies to the import of an 'all' export.

**Comment by dentoir** [comment]
Hi @brendam 

I hope the phrases import is working now in both scenarios. In the latest commit I've also changed the behaviour of the 'structure' export option. It should now create an empty bin during import; this will not cause the capture/ panel to report an error about missing tables.

**Comment by brendam** [comment]
Importing of bin from just 'structure' is working now - both with existing and new keywords.
Importing of bin from 'all' export also works.

So this is all good now for someone who wants to export and import a single bin at a time. 
Is it worth making these options available through the capture page instead of having to have command line access to use them?

Doesn't support multiple bins at once.

**Comment by hoylen** [comment]
I've modified the _export.php_ and _import.php_ scripts to handle multiple bins, as well as exporting every query bin (without having to explicitly specify their names).

Before submitting the changes, I wonder if someone could confirm this is the correct approach? (It seems to work in my testing, but testing cannot prove there are no problems.)

The current export script is something like:

```php
$bin = $argv[1];
// do export of $bin into the file called $filename
system("$bin_gzip $filename");
```

Is it correct to simply wrapped extra code around that, so it becomes something like:

```php
if (query bins are specified on the command line) {
    $queryBins = ... // populate an array from the command line
} else {
    $queryBins = getAllbins();
}
foreach ($queryBins as $bin) {
    // do export of $bin into the file called $filename (same code as in old script)
}
system("$bin_gzip $filename");
```

And the core part of importing of the exported data (queries only, or queries and data) is unchanged: simply running the SQL through the _mysql_ command.

P.S. This export feature could be useful via the Web pages, but command line tools are definitely still required. We want to automate the export process, rather than require a human to spend time manually clicking on things in a Web browser.

**Comment by dentoir** [comment]
Hi @hoylen,

What format does the export.php command-line take with multiple bins? Specifically: how do you define the difference between structure only and both data & structure. Will this happen per bin, or will you use the same export option for all bins?

This seems like a useful feature, indeed.

**Comment by hoylen** [comment]
In a single execution of the _export.php_ script, it only does "structure only" (-s) or "data and structure" (-d) for all the bins. Not per bin. Same export option applies to all the bins in a given execution.

I don't think there is much need to mix-and-match different exports modes in one run. If that is required, just run the script twice to get two separate export files.

Here is the command line syntax:

    Usage: export.php [options] {queryBins...}
    Options:
      -d       export query phrases AND data (default)
      -s       export query pharases only, with no data
      -o file  output file (default: automatically generated)
      -h       show this help message

Zero or more bins can be named. If zero bins are specified, that means to export all the bins.

The following examples show the different ways to express the same thing.

If there are exactly two bins named,and the second bin is called "all" or "structure", the command interprets that according to the old`export.php binName [all | structure]` syntax for backward compatibility. This only causes problems if you want to export two named bins and they are both strangely named "all" and "structure" (if only one of them had a special name, make it the first named bin and they will both be treated as bin names).

Also, _data and structure_ is the default, to reduce the risk of data loss. For example, if you deleted the database and then realised that you wanted to export the _structure only_, it is not a complete disaster. But if you had exported _structure only_ and then realised you needed _data and structure_, it is too late.

Export data and structure for all bins (i.e. everything):

    export.php

Export structure only for all bins:

    export.php -s

Export data and structure for a bin called "foo":

    export.php foo
    export.php -d foo
    export.php foo all

Export data and structure for a bin called "foo" and a bin called "bar":

    export.php foo bar
    export.php -d foo bar

Export structure only for a bin called "foo":

     export.php -s foo
     export.php foo structure

Export structure and data for a bin called "foo", a bin called "bar" and a bin called "baz":

    export.php foo bar baz
    export.php -d foo bar baz

Export structure only for a bin called "foo", a bin called "bar" and a bin called "baz":

    export.php -s foo bar baz

**Comment by dentoir** [comment]
Sounds good ;-)

Since these scripts are relatively new, I think we should remove the existing behaviour with the trailing 'structure' and 'all' and just use your flags (-d -s etc..).

Otherwise if we specify multiple bins on the command-line, there is an ambiguity when one of those bins is called 'structure' or 'all'.

**Comment by hoylen** [comment]
Ok, I'll remove the support for the legacy command line syntax. That makes it nice and clean.

So simply wrapping new code around the old single-bin-export code will generate the correct result?

I'll check it in after my existing pull request has been processed. It seems if I check in the changes now (into the HEAD of the master branch), they get included into the existing pull request.

**Comment by dentoir** [comment]
The scripts are there. Closing.
